overlord/snapstate/snapstate.go

```diff
@@ -134,7 +137,15 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 		stop := st.NewTask("stop-snap-services", fmt.Sprintf(i18n.G("Stop snap %q services"), snapsup.Name()))
 		addTask(stop)
 		prev = stop
+	}
+
+	if runRefreshHooks {
+		preRefreshHook := SetupPreRefreshHook(st, snapsup.Name())
+		addTask(preRefreshHook)
+		prev = preRefreshHook
+	}
 
+	if snapst.Active {
```

Comment:
Also, this sounds suspect. Does it mean we're running the hook even when the snap is not active? In which context would this make sense, or otherwise in which ways will it create problems?

Reply:
You're right about running the hook before stopping rather than after. I missed this aspect as I considered pre- and post- refresh hooks to be about running before and after switching active revision.